### ISSUE TITLE
GitHubService potentially sets wrong ref name on push

### ISSUE THREAD

**Issue by milo-minderbinder** [checkmarx-ltd/cx-flow#560]
While reading the code, I noticed that the line at https://github.com/checkmarx-ltd/cx-flow/blob/c05a60276e1e1903e2d934d72c00f2cc1d12e3fb/src/main/java/com/checkmarx/flow/service/GitHubService.java#L157 seems to set the ref name for push events incorrectly in the case where the branch name contains more than one forward-slash. For example, it’s fairly common to name branches with `feature/` or `bugfix/` as a prefix to denote the broad category the change list falls under. So, in the case where a branch is named `bugfix/master`, for example, the line linked above seems to incorrectly set the ref name to `master`, which is probably not intended. This issue still needs investigation though because I just noticed it while reading the code and have not tested it or observed the effect in a running environment.

As an aside, there are probably going to be other edge cases that pop up if the actual, full ref name is not kept verbatim with the request metadata as it goes through the processing flow, in GitHub or in other scm environments, because it’s actually very complicated to try to reverse engineer the short ref name for a given ref. For example, it’s totally valid to have a branch called refs/heads/main, and also have a tag called refs/tags/main, in which case “main” is an ambiguous/invalid short name. You could try to call the GitHub git/matching-refs endpoint to check, or locally shell out to git to call, eg, `git rev-parse —abbrev-ref` to get the definitive unambiguous short name.but it’s probably easier and less error prone to keep the full ref name and then strip the leading refs/heads/ before calling out to Checkmarx APIs.. though I think those technically support branch names and tags so...

**Comment by AsafCX** [comment]
Going over the entire flow you may notice the mentioned code used only to extract branch name, ref name remains as-is.
you may also notice this note above the code:
_**// According to GitHub the recommended way to extract the branch name
// is by using the 'ref' parameter which is in the following format: 'refs/heads/<branch>'**_
Meaning, ref parameter remains the same only using it for branch parameter
Github prevents the usage of '/' as part of the branch name so the case in which the actual branch name contains '/' can't happen.

**Comment by milo-minderbinder** [comment]
@AsafCX I'll try to clarify because I don't think you've understood what I meant.

> Github prevents the usage of '/' as part of the branch name so the case in which the actual branch name contains '/' can't happen.
This is definitely incorrect. In fact, if you fetch the refs for this repo, there are currently branches with names containing a `/`:
```shell
$> git ls-remote git@github.com:checkmarx-ltd/cx-flow.git
e15879960b6892585bead904f9bdb69682198184	HEAD
e15879960b6892585bead904f9bdb69682198184	refs/heads/develop
0420345e5f58a1382114d82494fa347e8dceb1a5	refs/heads/fix/capone-shard-fix
6d89856137af029fc19a76f569901afdf268f29a	refs/heads/master
b2a2108caa43b808aa7061294b25a11896691a04	refs/heads/pr-nimrod-support-bitbucket-e2e
77897fd8957910b8cd8a95ab648742fe0a141b96	refs/pull/1/head
<--...output truncated...-->
51942b64838efa76c6d04ca3fdf56ce87483f7d1	refs/tags/ts-1.3^{}
ba6b73992426585250ecdf8d7ad76d49ac16e0bf	refs/tags/v1.6.4
```
As you can see, the branch `fix/capone-shard-fix` contains a forward-slash.

And _because_ `fix/capone-shard-fix` contains a forward-slash, line 157 linked above will incorrectly set the `branch` parameter in the `RepoReader` constructor to `capone-shard-fix` since this part strips everything up to the _last_ forward-slash: `ref.substring(ref.lastIndexOf('/') + 1)`.

A simple fix would just be to do:
```diff
diff --git a/src/main/java/com/checkmarx/flow/service/GitHubService.java b/src/main/java/com/checkmarx/flow/service/GitHubService.java
index 312cb577..e474e762 100644
--- a/src/main/java/com/checkmarx/flow/service/GitHubService.java
+++ b/src/main/java/com/checkmarx/flow/service/GitHubService.java
@@ -155,7 +155,7 @@ public class GitHubService extends RepoService {
                 // According to GitHub the recommended way to extract the branch name
                 // is by using the 'ref' parameter which is in the following format: 'refs/heads/<branch>'
                 configProvider.init(uid, new RepoReader(properties.getApiUrl(), repository.getOwner().getName(),
-                        repository.getName(), ref.substring(ref.lastIndexOf('/') + 1),
+                        repository.getName(), ref.replaceFirst("^refs/heads/", ""),
                         properties.getToken(), SourceProviderType.GITHUB));
             } catch (ConfigurationException e) {
                 log.warn("Failed to init config provider with the following error: {}", e.getMessage());
```
That would simply strip the `refs/heads/` part from the beginning of the full ref name, which is what I think was intended in the first place.

**Comment by AsafCX** [comment]
You are right, Branch name extraction in GitHubService is wrong.
We already support and use the right extraction under ScanUtlis , I'll update it.
thanks for the warning

**Comment by AvivCx** [comment]
fixed in cxflow 1.6.17
